### PR TITLE
Gimbal rebased for master

### PR DESCRIPTION
**Pull request by arthurbenemann**
Changes required for the rate-controlled gimbal.

@tridge I just did the rebase on master, and have not tested the build. I'll do a build tomorrow when I'm back at my workstation, and have hardware to test with.

### CODE REVIEW

File: ArduCopter/ArduCopter.pde [hunk 1]
```diff
@@ -132,6 +132,7 @@
 #include <AP_ServoRelayEvents.h>
 #include <AP_Camera.h>          // Photo or video camera
 #include <AP_Mount.h>           // Camera/Antenna mount
+#include <AP_Gimbal.h>          // Camera rate controlled gimbal
```
Comment (arthurbenemann) [inline ArduCopter/ArduCopter.pde]: probably not required anymore

File: ArduCopter/ArduCopter.pde [hunk 2]
```diff
@@ -743,7 +744,7 @@ static const AP_Scheduler::Task scheduler_tasks[] PROGMEM = {
     { ekf_check,            40,      2 },
     { crash_check,          40,      2 },
     { landinggear_update,   40,      1 },
-    { gcs_check_input,	     8,    550 },
+    { gcs_check_input,	     1,    550 },
```
Comment (arthurbenemann) [inline ArduCopter/ArduCopter.pde]: this is the most concerning change for the flight code. But it's required to ensure that the rate loop on the gimbal run's at 100Hz without too much latency between report and reply msgs form the gimbal

File: ArduCopter/GCS_Mavlink.pde [hunk 3]
```diff
@@ -990,6 +990,8 @@ void GCS_MAVLINK::handleMessage(mavlink_message_t* msg)
 #if MOUNT == ENABLED
         handle_gimbal_report(camera_mount, msg);
 #endif
+        //gimbal.receive_feedback(chan, msg);
```
Comment (arthurbenemann) [inline ArduCopter/GCS_Mavlink.pde]: clean-up needed

File: ArduCopter/GCS_Mavlink.pde [hunk 4]
```diff
@@ -990,6 +990,8 @@ void GCS_MAVLINK::handleMessage(mavlink_message_t* msg)
...
+        //Log_Write_Gimbal();
```
Comment (arthurbenemann) [inline ArduCopter/GCS_Mavlink.pde]: looging needs to be re-implemented

File: libraries/AP_Mount/AP_Gimbal.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
+// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+
+/************************************************************
+* AP_Gimbal -- library to control a 3 axis rate gimbal.		*
+*															*
+* Author:  Arthur Benemann, Paul Riseborough;									*
+*															*
```
Comment (arthurbenemann) [inline libraries/AP_Mount/AP_Gimbal.h]: some extra info would be nice

File: libraries/AP_Mount/AP_Mount.cpp [hunk 6]
```diff
@@ -199,6 +199,11 @@ const AP_Param::GroupInfo AP_Mount::var_info[] PROGMEM = {
     // @User: Standard
     AP_GROUPINFO("_TYPE", 19, AP_Mount, state[0]._type, 0),
 
+    AP_GROUPINFO("_OFF_JNT", 20, AP_Mount, state[0]._gimbalParams.joint_angles_offsets, 0),
```
Comment (arthurbenemann) [inline libraries/AP_Mount/AP_Mount.cpp]: The offsets will probably need to be moved to the gimbal itself. SO the units are swap-able without requiring re-calibration ( the gimbal keeps it's calibration parameters)

File: libraries/AP_Mount/AP_Mount.h [hunk 7]
```diff
@@ -65,6 +65,13 @@ class AP_Mount
         Mount_Type_SToRM32 = 4          /// SToRM32 mount
     };
 
+    struct gimbal_params {
```
Comment (arthurbenemann) [inline libraries/AP_Mount/AP_Mount.h]: maybe move these to the AP_GImbal class

File: libraries/GCS_MAVLink/message_definitions/ardupilotmega.xml [hunk 8]
```diff
@@ -555,8 +650,134 @@
         <field type="float" name="demanded_rate_z">Demanded angular rate Z (rad/s)</field>
     </message>
 
-    <!-- 202 to 214 RESERVED for gimbal control -->
-    <!-- 215 to 218 RESERVED for camera control -->
+    <message name="GIMBAL_RESET" id="202">
```
Comment (arthurbenemann) [inline libraries/GCS_MAVLink/message_definitions/ardupilotmega.xml]: The extra gimbal msgs can probably be collapsed into a smaller set

File: libraries/GCS_MAVLink/message_definitions/ardupilotmega.xml [hunk 9]
```diff
@@ -555,8 +650,134 @@
...
+        <field name="test" type="uint8_t" enum="FACTORY_TEST">Which factory test is currently running</field>
+        <field name="test_section" type="uint8_t">Which section of the test is currently running.  The meaning of this is test-dependent</field>
+        <field name="test_section_progress" type="uint8_t">The progress of the current test section, 0x64=100%</field>
+        <field name="test_status" type="uint8_t">The status of the currently executing test section.  The meaning of this is test and section-dependent</field>
+    </message>
+
+    <!-- 211 to 214 RESERVED for more GIMBAL -->
+
+    <message name="GOPRO_POWER_ON" id="215">
+        <description>Instruct a HeroBus attached GoPro to power on</description>
```
Comment (arthurbenemann) [inline libraries/GCS_MAVLink/message_definitions/ardupilotmega.xml]: gopro messages are likely to change

Comment (arthurbenemann) [inline libraries/GCS_MAVLink/message_definitions/ardupilotmega.xml]: Gopro messages are closer to final now. But they can still be collapsed, and possible be using commandlong messages instead.

**Comment by gmorph** [comment]
Hi Arthur.  Travis has now been fixed and this PR is not passing Travis.  Can you have a look please?  Thanks.

**Comment by rmackay9** [comment]
$20 says it breaks the build for plane, rover and antenna tracker..

**Comment by rmackay9** [comment]
Arthur, thanks a lot for this commit.  I've created a modified version of the PR (https://github.com/rmackay9/rmackay9-ardupilot/commits/arthur-gimbal2) with some issues resolved.
1. fixed build so it now passes travis (it was breaking for Rover/Plane on APM2)
2. added parameter descriptions for _OFF_GYRO, _OFF_ACC, _K_RATE, _OFF_JNT

When I've loaded this onto my solo gimbal equipped vehicle neither your original commits nor my modified ones cause the gimbal to move.  The latest solo version from Jonathan does work though, any ideas what's wrong?

There are some other questions/issues outstanding:
1. are the new parameter descriptions correct?
2. Do these _OFF_* parameters really need to be saved across vehicle reboots?
3. isCopterFlipped() should probably be renamed to isVehicleFlipped.
4. there's a mixture of CamelCase and underscore.  i.e. send_control and isCopterFlipped use different conventions.  We should at least be consistent within a library and ideally use the project default which is to use underscores.
5. we should consider merging the AP_Gimbal class into the AP_Mount_MAVLink class and perhaps rename AP_Mount_MAVLink to AP_Mount_Solo.

**Comment by arthurbenemann** [comment]
When I've loaded this onto my solo gimbal equipped vehicle neither your original commits nor my modified ones cause the gimbal to move. The latest solo version from Jonathan does work though, any ideas what's wrong?The latest is 0.1.3, and was released by me ( 0.1.2 has some issues when using the gimbal). The issues are likely the parameter setting, or differences in the way the two branches are handling serial ports.

1. are the new parameter descriptions correct?
The ones on the ROMFS are.

2. Do these OFF* parameters really need to be saved across vehicle reboots?
They do need to be saved. But they will all be moved to the gimbal itself in a near future, so we swap gimbals between copters without re-calibraton.

3. isCopterFlipped() should probably be renamed to isVehicleFlipped.
No problem. And it's more like isVehicleOrientationOutOfGimbalOperationRange(), but that is a bit long.

4. there's a mixture of CamelCase and underscore. i.e. send_control and isCopterFlipped use different conventions. We should at least be consistent within a library and ideally use the project default which is to use underscores.
The result of different moods, I'll follow the project defaults next time. Do we have them written somewhere?

5. we should consider merging the AP_Gimbal class into the AP_Mount_MAVLink class and perhaps rename AP_Mount_MAVLink to AP_Mount_Solo.
Renaming to Mount_solo makes sense. But I stil think the code should be in a separate file, as it's dealing with controlling the gimbal to point somewhere. And not deciding where to point it to, or handling a vehicle state machine (GPS,RC, ...). It a future revision this file can be run in a uC inside the gimbal itself.

**Comment by rmackay9** [comment]
Ah, right so perhaps it's not working because the appropriate serial port's baud rate is incorrect.  Any idea what baud rate the gimbal uses?

**Comment by rmackay9** [comment]
Ok, so from loading up the working version and checking it's params I've got it basically working.
For the record it needed:
SERIAL3_PROTOCOL=0  <-- to disable the serial3
SERIAL4_PROTOCOL=1 <-- so serial 4 is mavlink protocol
SERIAL4_BAUD = 230
the gimbal moves now although it seems to suffer from drift in the roll axis which I didn' see with solo-master.

**Comment by rmackay9** [comment]
merged.  txs!

**Comment by arthurbenemann** [comment]
Thanks for merging.

I'll give a try in the master branch when I have some time.

Let me know when you want to update your gimbal. I hope you left with a version that has a working bootloader.

**Pull request closed without merging**